File: backend/app/storage.py
"""httpx と AWS SigV4 を使用した S3 互換ストレージクライアント。"""

import hashlib
import hmac
from collections.abc import AsyncIterator
from datetime import datetime, timezone
from urllib.parse import quote, urlparse

import httpx

from app.config import settings

# M-14: S3操作用の共有HTTPクライアント
_s3_client: httpx.AsyncClient | None = None


def _get_s3_client() -> httpx.AsyncClient:
    global _s3_client
    if _s3_client is None or _s3_client.is_closed:
        from app.utils.http_client import USER_AGENT
        _s3_client = httpx.AsyncClient(
            base_url=settings.s3_endpoint_url,
            timeout=30.0,
            headers={"User-Agent": USER_AGENT},
        )
    return _s3_client


def _sign(key: bytes, msg: str) -> bytes:
    return hmac.new(key, msg.encode(), hashlib.sha256).digest()


def _signing_key(date_str: str) -> bytes:
    k = _sign(("AWS4" + settings.s3_secret_access_key).encode(), date_str)
    k = _sign(k, settings.s3_region)
    k = _sign(k, "s3")
    return _sign(k, "aws4_request")


def _endpoint_host() -> str:
    return urlparse(settings.s3_endpoint_url).netloc


def _auth_headers(
    method: str,
    path: str,
    content_sha256: str,
    extra_headers: dict[str, str] | None = None,
) -> dict[str, str]:
    now = datetime.now(timezone.utc)
    date_str = now.strftime("%Y%m%d")
    amz_date = now.strftime("%Y%m%dT%H%M%SZ")

    headers = {
        "host": _endpoint_host(),
        "x-amz-date": amz_date,
        "x-amz-content-sha256": content_sha256,
    }
    if extra_headers:
        headers.update(extra_headers)

    signed_headers_list = sorted(headers.keys())
    signed_headers_str = ";".join(signed_headers_list)
    canonical_headers = "".join(f"{k}:{headers[k]}\n" for k in signed_headers_list)

    canonical_request = "\n".join(
        [
            method,
            quote(path, safe="/"),
            "",
            canonical_headers,
            signed_headers_str,
            content_sha256,
        ]
    )

    credential_scope = f"{date_str}/{settings.s3_region}/s3/aws4_request"
    string_to_sign = "\n".join(
        [
            "AWS4-HMAC-SHA256",
            amz_date,
            credential_scope,
            hashlib.sha256(canonical_request.encode()).hexdigest(),
        ]
    )

    signature = hmac.new(
        _signing_key(date_str),
        string_to_sign.encode(),
        hashlib.sha256,
    ).hexdigest()

    headers["Authorization"] = (
        f"AWS4-HMAC-SHA256 "
        f"Credential={settings.s3_access_key_id}/{credential_scope}, "
        f"SignedHeaders={signed_headers_str}, "
        f"Signature={signature}"
    )
    return headers


async def ensure_bucket() -> None:
    """バケットが存在しない場合に作成する。"""
    path = f"/{settings.s3_bucket}"
    content_sha256 = hashlib.sha256(b"").hexdigest()
    headers = _auth_headers("PUT", path, content_sha256)
    client = _get_s3_client()
    resp = await client.put(path, headers=headers)
    if resp.status_code not in (200, 409):
        resp.raise_for_status()


async def upload_file(key: str, data: bytes, content_type: str) -> str:
    """ファイルを S3 にアップロードする。ETag を返す。"""
    path = f"/{settings.s3_bucket}/{key}"
    content_sha256 = hashlib.sha256(data).hexdigest()
    extra = {"content-type": content_type}
    headers = _auth_headers("PUT", path, content_sha256, extra)
    headers["content-type"] = content_type
    client = _get_s3_client()
    resp = await client.put(path, content=data, headers=headers)
    resp.raise_for_status()
    return resp.headers.get("etag", "").strip('"')


async def delete_file(key: str) -> None:
    """S3 からファイルを削除する。"""
    path = f"/{settings.s3_bucket}/{key}"
    content_sha256 = hashlib.sha256(b"").hexdigest()
    headers = _auth_headers("DELETE", path, content_sha256)
    client = _get_s3_client()
    resp = await client.delete(path, headers=headers)
    if resp.status_code not in (200, 204, 404):
        resp.raise_for_status()


async def download_file(key: str) -> bytes:
    """S3 からファイルをダウンロードし、その内容を返す。"""
    path = f"/{settings.s3_bucket}/{key}"
    content_sha256 = "UNSIGNED-PAYLOAD"
    headers = _auth_headers("GET", path, content_sha256)
    client = _get_s3_client()
    resp = await client.get(path, headers=headers)
    resp.raise_for_status()
    return resp.content


async def get_file_stream(key: str) -> tuple[AsyncIterator[bytes], str, int]:
    """S3 からファイルをストリームとして取得する。(stream, content_type, size) を返す。"""
    path = f"/{settings.s3_bucket}/{key}"
    content_sha256 = "UNSIGNED-PAYLOAD"
    headers = _auth_headers("GET", path, content_sha256)
    client = _get_s3_client()
    resp = await client.send(
        client.build_request("GET", path, headers=headers),
        stream=True,
    )
    resp.raise_for_status()
    ct = resp.headers.get("content-type", "application/octet-stream")
    size = int(resp.headers.get("content-length", 0))

    async def _stream():
        try:
            async for chunk in resp.aiter_bytes():
                yield chunk
        finally:
            await resp.aclose()
            await client.aclose()

    return _stream(), ct, size


def get_public_url(key: str) -> str:
    """指定された S3 キーの公開 URL を返す。"""
    return f"{settings.media_url}/{key}"


def generate_presigned_get_url(key: str, expires_in: int = 300) -> str:
    """S3 オブジェクトの GET 用 presigned URL を返す (AWS SigV4 query string 方式)。

    video-thumb の `/thumbnail_from_url` 等、内部マイクロサービスから一時的に
    S3 オブジェクトを取得させたい場面で使う。

    Args:
        key: S3 object key (bucket prefix なし)
        expires_in: 有効期限 (秒、1 〜 604800 = 7 日)

    Returns:
        presigned URL (`{s3_endpoint_url}/{bucket}/{key}?X-Amz-Algorithm=...`)

    Raises:
        ValueError: expires_in が AWS 仕様の範囲 (1-604800 秒) を超える場合
    """
    if not 1 <= expires_in <= 604800:
        raise ValueError(
            f"expires_in must be between 1 and 604800 seconds, got {expires_in}"
        )
    now = datetime.now(timezone.utc)
    date_str = now.strftime("%Y%m%d")
    amz_date = now.strftime("%Y%m%dT%H%M%SZ")
    credential_scope = f"{date_str}/{settings.s3_region}/s3/aws4_request"

    host = _endpoint_host()
    path = f"/{settings.s3_bucket}/{key}"
    canonical_uri = quote(path, safe="/")

    # 署名対象は host ヘッダのみ (presigned URL の慣習)
    signed_headers = "host"
    canonical_headers = f"host:{host}\n"

    # クエリ文字列 (キー名はソートする必要がある)
    query_params = {
        "X-Amz-Algorithm": "AWS4-HMAC-SHA256",
        "X-Amz-Credential": f"{settings.s3_access_key_id}/{credential_scope}",
        "X-Amz-Date": amz_date,
        "X-Amz-Expires": str(expires_in),
        "X-Amz-SignedHeaders": signed_headers,
    }
    canonical_query = "&".join(
        f"{quote(k, safe='-_.~')}={quote(v, safe='-_.~')}"
        for k, v in sorted(query_params.items())
    )

    canonical_request = "\n".join(
        [
            "GET",
            canonical_uri,
            canonical_query,
            canonical_headers,
            signed_headers,
            "UNSIGNED-PAYLOAD",
        ]
    )
    string_to_sign = "\n".join(
        [
            "AWS4-HMAC-SHA256",
            amz_date,
            credential_scope,
            hashlib.sha256(canonical_request.encode()).hexdigest(),
        ]
    )
    signature = hmac.new(
        _signing_key(date_str),
        string_to_sign.encode(),
        hashlib.sha256,
    ).hexdigest()

    return (
        f"{settings.s3_endpoint_url}{canonical_uri}"
        f"?{canonical_query}&X-Amz-Signature={signature}"
    )

File: backend/backend-rs/src/storage.rs
//! `app/storage.py` のうち `ensure_bucket`/`upload_file`/`delete_file`/
//! `download_file` を移植したもの。手書き AWS SigV4 (boto3 非依存) を
//! `reqwest` で踏襲する。`get_file_stream`/`generate_presigned_get_url` は
//! まだ backend-rs 側に呼び出し元が無いため未移植(`download_file`はemoji
//! エクスポート(#1139 Stage 4、`admin.rs`の`export_emojis_endpoint`)が
//! ZIPに書き出す前に画像を丸ごとメモリに読む用途のみで、ストリーミングは
//! 要しない)。

use std::sync::OnceLock;
use std::time::Duration;

use axum::http::StatusCode;
use hmac::{Hmac, Mac};
use sha2::{Digest, Sha256};

use crate::config::Config;
use crate::error::AppError;

type HmacSha256 = Hmac<Sha256>;

static S3_CLIENT: OnceLock<reqwest::Client> = OnceLock::new();

/// `app.storage._get_s3_client`相当のプロセス内共有クライアント。
fn http_client() -> &'static reqwest::Client {
    S3_CLIENT.get_or_init(|| {
        reqwest::Client::builder()
            .timeout(Duration::from_secs(30))
            .build()
            .expect("failed to build S3 http client")
    })
}

fn hmac_sha256(key: &[u8], msg: &str) -> Vec<u8> {
    let mut mac = HmacSha256::new_from_slice(key).expect("HMAC accepts a key of any length");
    mac.update(msg.as_bytes());
    mac.finalize().into_bytes().to_vec()
}

/// `app.storage._signing_key`を移植したもの。
fn signing_key(config: &Config, date_str: &str) -> Vec<u8> {
    let k = hmac_sha256(
        format!("AWS4{}", config.s3_secret_access_key).as_bytes(),
        date_str,
    );
    let k = hmac_sha256(&k, &config.s3_region);
    let k = hmac_sha256(&k, "s3");
    hmac_sha256(&k, "aws4_request")
}

/// `app.storage._endpoint_host`を移植したもの (`urlparse(...).netloc`相当)。
fn endpoint_host(config: &Config) -> Result<String, AppError> {
    let url = reqwest::Url::parse(&config.s3_endpoint_url)
        .map_err(|_| AppError::new(StatusCode::INTERNAL_SERVER_ERROR, "invalid S3 endpoint URL"))?;
    let host = url.host_str().ok_or_else(|| {
        AppError::new(StatusCode::INTERNAL_SERVER_ERROR, "invalid S3 endpoint URL")
    })?;
    Ok(match url.port() {
        Some(port) => format!("{host}:{port}"),
        None => host.to_string(),
    })
}

/// AWS SigV4 の canonical URI エンコード (`urllib.parse.quote(path, safe="/")`相当)。
/// 未予約文字 (`A-Za-z0-9-._~`) と `/` 以外を `%XX` にパーセントエンコードする。
fn uri_encode_path(path: &str) -> String {
    let mut out = String::with_capacity(path.len());
    for b in path.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'.' | b'_' | b'~' | b'/' => {
                out.push(b as char)
            }
            _ => out.push_str(&format!("%{b:02X}")),
        }
    }
    out
}

/// `app.storage._auth_headers`を移植したもの。署名済みヘッダー一覧
/// (`Authorization`込み、ソート済み)を返す。
fn auth_headers(
    config: &Config,
    method: &str,
    encoded_path: &str,
    content_sha256: &str,
    extra_headers: &[(&str, String)],
) -> Result<Vec<(String, String)>, AppError> {
    let now = chrono::Utc::now();
    let date_str = now.format("%Y%m%d").to_string();
    let amz_date = now.format("%Y%m%dT%H%M%SZ").to_string();

    let mut headers: Vec<(String, String)> = vec![
        ("host".to_string(), endpoint_host(config)?),
        ("x-amz-date".to_string(), amz_date.clone()),
        (
            "x-amz-content-sha256".to_string(),
            content_sha256.to_string(),
        ),
    ];
    for (k, v) in extra_headers {
        headers.push((k.to_string(), v.clone()));
    }
    headers.sort_by(|a, b| a.0.cmp(&b.0));

    let signed_headers_str = headers
        .iter()
        .map(|(k, _)| k.as_str())
        .collect::<Vec<_>>()
        .join(";");
    let canonical_headers: String = headers.iter().map(|(k, v)| format!("{k}:{v}\n")).collect();

    let canonical_request = [
        method,
        encoded_path,
        "", // canonical query string (このクライアントは常に空)
        &canonical_headers,
        &signed_headers_str,
        content_sha256,
    ]
    .join("\n");

    let credential_scope = format!("{date_str}/{}/s3/aws4_request", config.s3_region);
    let string_to_sign = format!(
        "AWS4-HMAC-SHA256\n{amz_date}\n{credential_scope}\n{:x}",
        Sha256::digest(canonical_request.as_bytes())
    );

    let signature_bytes = {
        let mut mac = HmacSha256::new_from_slice(&signing_key(config, &date_str))
            .expect("HMAC accepts a key of any length");
        mac.update(string_to_sign.as_bytes());
        mac.finalize().into_bytes()
    };
    let signature = format!("{signature_bytes:x}");

    headers.push((
        "Authorization".to_string(),
        format!(
            "AWS4-HMAC-SHA256 Credential={}/{credential_scope}, SignedHeaders={signed_headers_str}, Signature={signature}",
            config.s3_access_key_id
        ),
    ));
    Ok(headers)
}

fn request_error(err: reqwest::Error) -> AppError {
    tracing::error!(error = %err, "S3 request failed");
    AppError::new(StatusCode::INTERNAL_SERVER_ERROR, "Internal server error")
}

fn build_url(config: &Config, encoded_path: &str) -> String {
    format!(
        "{}{encoded_path}",
        config.s3_endpoint_url.trim_end_matches('/')
    )
}

/// `app.storage.ensure_bucket`を移植したもの。バケットが既に存在する場合の
/// `409`もPython版と同じく正常系として扱う。
pub async fn ensure_bucket(config: &Config) -> Result<(), AppError> {
    let path = format!("/{}", config.s3_bucket);
    let encoded_path = uri_encode_path(&path);
    let content_sha256 = format!("{:x}", Sha256::digest(b""));
    let headers = auth_headers(config, "PUT", &encoded_path, &content_sha256, &[])?;

    let mut req = http_client().put(build_url(config, &encoded_path));
    for (k, v) in &headers {
        req = req.header(k.as_str(), v.as_str());
    }
    let resp = req.send().await.map_err(request_error)?;
    if resp.status() != StatusCode::OK && resp.status() != StatusCode::CONFLICT {
        tracing::warn!(status = %resp.status(), "S3 ensure_bucket returned unexpected status");
    }
    Ok(())
}

/// `app.storage.upload_file`を移植したもの。
pub async fn upload_file(
    config: &Config,
    key: &str,
    data: &[u8],
    content_type: &str,
) -> Result<(), AppError> {
    let path = format!("/{}/{key}", config.s3_bucket);
    let encoded_path = uri_encode_path(&path);
    let content_sha256 = format!("{:x}", Sha256::digest(data));
    let headers = auth_headers(
        config,
        "PUT",
        &encoded_path,
        &content_sha256,
        &[("content-type", content_type.to_string())],
    )?;

    let mut req = http_client()
        .put(build_url(config, &encoded_path))
        .body(data.to_vec());
    for (k, v) in &headers {
        req = req.header(k.as_str(), v.as_str());
    }
    let resp = req.send().await.map_err(request_error)?;
    if !resp.status().is_success() {
        tracing::error!(status = %resp.status(), key, "S3 upload failed");
        return Err(AppError::new(
            StatusCode::INTERNAL_SERVER_ERROR,
            "Internal server error",
        ));
    }
    Ok(())
}

/// `app.storage.delete_file`を移植したもの。`404`(既に存在しない)も
/// Python版と同じく正常系として扱う。
pub async fn delete_file(config: &Config, key: &str) -> Result<(), AppError> {
    let path = format!("/{}/{key}", config.s3_bucket);
    let encoded_path = uri_encode_path(&path);
    let content_sha256 = format!("{:x}", Sha256::digest(b""));
    let headers = auth_headers(config, "DELETE", &encoded_path, &content_sha256, &[])?;

    let mut req = http_client().delete(build_url(config, &encoded_path));
    for (k, v) in &headers {
        req = req.header(k.as_str(), v.as_str());
    }
    let resp = req.send().await.map_err(request_error)?;
    let status = resp.status();
    if status != StatusCode::OK
        && status != StatusCode::NO_CONTENT
        && status != StatusCode::NOT_FOUND
    {
        tracing::error!(status = %status, key, "S3 delete failed");
        return Err(AppError::new(
            StatusCode::INTERNAL_SERVER_ERROR,
            "Internal server error",
        ));
    }
    Ok(())
}

/// `app.storage.get_public_url`を移植したもの。
pub fn public_url(config: &Config, key: &str) -> String {
    format!("{}/{key}", config.media_url())
}

/// `app.storage.download_file`を移植したもの。`404`は`None`として返す
/// (Python版は`resp.raise_for_status()`で例外化するが、呼び出し元の
/// `export_emojis_endpoint`は「S3上に実体が無ければそのカスタム絵文字を
/// ZIPから静かに除外する」ベストエフォート処理のため、backend-rs側は
/// エラーと未検出を呼び出し元で区別できるよう`Option`にした)。
/// Python版の`download_file`/`get_file_stream`と同じく`content_sha256`に
/// 実ハッシュではなく`"UNSIGNED-PAYLOAD"`を使う(GETはボディが無いため
/// `upload_file`/`delete_file`の空文字列ハッシュと等価だが、Python版の
/// 実際の挙動をそのまま踏襲する)。
pub async fn get_file(config: &Config, key: &str) -> Result<Option<Vec<u8>>, AppError> {
    let path = format!("/{}/{key}", config.s3_bucket);
    let encoded_path = uri_encode_path(&path);
    let headers = auth_headers(config, "GET", &encoded_path, "UNSIGNED-PAYLOAD", &[])?;

    let mut req = http_client().get(build_url(config, &encoded_path));
    for (k, v) in &headers {
        req = req.header(k.as_str(), v.as_str());
    }
    let resp = req.send().await.map_err(request_error)?;
    if resp.status() == StatusCode::NOT_FOUND {
        return Ok(None);
    }
    if !resp.status().is_success() {
        tracing::error!(status = %resp.status(), key, "S3 get failed");
        return Err(AppError::new(
            StatusCode::INTERNAL_SERVER_ERROR,
            "Internal server error",
        ));
    }
    let bytes = resp.bytes().await.map_err(request_error)?;
    Ok(Some(bytes.to_vec()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn uri_encode_path_keeps_slashes_and_unreserved_chars() {
        assert_eq!(
            uri_encode_path("/bucket/server/abc-123.png"),
            "/bucket/server/abc-123.png"
        );
    }

    #[test]
    fn uri_encode_path_escapes_reserved_chars() {
        assert_eq!(uri_encode_path("/bucket/a b+c"), "/bucket/a%20b%2Bc");
    }
}
